controller api failed while doing parallel IO with error :java.util.concurrent.CompletionException: io.pravega.controller.store.stream.StoreException$DataNotFoundException

Problem description:
 Error was found in controller log after running parallel IO with the following specifications:
```


2019-04-18 07:55:52,803 82767 [controllerpool-8] ERROR i.p.c.s.r.g.v.ControllerServiceImpl - Controller api failed with error:
java.util.concurrent.CompletionException: io.pravega.controller.store.stream.StoreException$DataNotFoundException: get entry: key: _system table: _system/_tables/scopes
        at io.pravega.controller.store.stream.PravegaTablesStoreHelper.lambda$null$49(PravegaTablesStoreHelper.java:515)
        at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:870)
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:852)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
        at io.pravega.controller.server.SegmentHelper$14.noSuchSegment(SegmentHelper.java:973)
        at io.pravega.controller.server.SegmentStoreConnectionManager$ReusableReplyProcessor.execute(SegmentStoreConnectionManager.java:235)
        at io.pravega.controller.server.SegmentStoreConnectionManager$ReusableReplyProcessor.noSuchSegment(SegmentStoreConnectionManager.java:273)
        at io.pravega.shared.protocol.netty.WireCommands$NoSuchSegment.process(WireCommands.java:279)
        at io.pravega.shared.protocol.netty.ReplyProcessor.process(ReplyProcessor.java:20)
        at io.pravega.client.netty.impl.FlowHandler.lambda$channelRead$2(FlowHandler.java:195)
        at java.util.Optional.ifPresent(Optional.java:159)
        at io.pravega.client.netty.impl.FlowHandler.channelRead(FlowHandler.java:193)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
        at io.pravega.shared.protocol.netty.ExceptionLoggingHandler.channelRead(ExceptionLoggingHandler.java:37)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
Caused by: io.pravega.controller.store.stream.StoreException$DataNotFoundException: get entry: key: _system table: _system/_tables/scopes
        at io.pravega.controller.store.stream.StoreException.create(StoreException.java:90)
        at io.pravega.controller.store.stream.PravegaTablesStoreHelper.lambda$null$49(PravegaTablesStoreHelper.java:493)
        ... 40 common frames omitted
Caused by: io.pravega.controller.server.WireCommandFailedException: WireCommandFailed with type READ_TABLE reason SegmentDoesNotExist
        ... 36 common frames omitted
2019-04-18 07:55:52,828 82792 [grpc-default-executor-0] INFO  i.p.c.s.r.g.v.ControllerServiceImpl - getCurrentSegments called for stream _system/_requeststream.
2019-04-18 07:55:52,832 82796 [epollEventLoopGroup-2-1] WARN  i.p.controller.server.SegmentHelper - [requestId=8] readTable _system/_tables/scopes NoSuchSegment
2019-04-18 07:55:52,833 82797 [controllerpool-11] ERROR i.p.c.s.r.g.v.ControllerServiceImpl - Controller api failed with error:
java.util.concurrent.CompletionException: io.pravega.controller.store.stream.StoreException$DataNotFoundException: get entry: key: _system table: _system/_tables/scopes
        at io.pravega.controller.store.stream.PravegaTablesStoreHelper.lambda$null$49(PravegaTablesStoreHelper.java:515)
```
Deployment Details:-
```
bookkeeper:0.5.0-2222.8ba5431
pravega:0.5.0-2222.8ba5431
zookeeper:3.5.4-beta-operator
pravega-operator:0.3.2
zookeeper-operator:0.2.1
```
Environment details: PKS / K8 with medium cluster:
```
3 master: xlarge: 4 CPU, 16 GB Ram, 32 GB Disk
5 worker: 2xlarge: 8 CPU, 32 GB Ram, 64 GB Disk
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend
```

Ran parallel i/o workload with follwoing configurations :
 ```
 IO 1:
 (Total -3 writers, events:4,858,891, events/sec:41, KB/sec:32.086914)
 Total -4 readers , events:34,337,261, events/sec:295, KB/sec:226.75488
 
 IO 2 :
 
 (Total -1 writers, events:157,457, events/sec:12, KB/sec:9.581055)
 Total -1 readers ,events:157,457, events/sec:12, KB/sec:9.581055 )
```
2.did not see any pod restart at that time 

3.saw the below error while doing IO in IO client:
```
ERROR i.p.l.t.w.w.PravegaWriterState - Error with BeginTransaction
java.lang.RuntimeException: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.grpc.StatusRuntimeException: INTERNAL: java.util.concurrent.CompletionException: io.pravega.controller.server.WireCommandFailedException: WireCommandFailed with type CREATE_SEGMENT reason ConnectionDropped
        at io.pravega.common.concurrent.Futures.getAndHandleExceptions(Futures.java:180)
        at io.pravega.client.stream.impl.EventStreamWriterImpl.beginTxn(EventStreamWriterImpl.java:315)
        at io.pravega.longevity.testworker.workers.writers.PravegaWriterState.beforeWrite(PravegaWriterState.kt:39)
        at io.pravega.longevity.testworker.workers.writers.PravegaWriter.start(PravegaWriter.java:143)
        at io.pravega.longevity.testworker.workers.writers.WriterWorkerManager.lambda$start$0(WriterWorkerManager.java:134)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.grpc.StatusRuntimeException: INTERNAL: java.util.concurrent.CompletionException: io.pravega.controller.server.WireCommandFailedException: WireCommandFailed with type CREATE_SEGMENT reason ConnectionDropped
```


4.root@osboxes:/home/osboxes/work/longivity# kubectl get pods --all-namespaces
````
NAMESPACE           NAME                                                 READY     STATUS      RESTARTS   AGE
default             pravega-benchmark                                    1/1       Running     0          4d
default             pravega-bookie-0                                     1/1       Running     0          4d
default             pravega-bookie-1                                     1/1       Running     0          4d
default             pravega-bookie-2                                     1/1       Running     1          4d
default             pravega-bookie-3                                     1/1       Running     0          4d
default             pravega-operator-779879b48-8dnf2                     1/1       Running     0          4d
default             pravega-pravega-controller-84948d8895-lknhc          1/1       Running     2          4d
default             pravega-pravega-controller-84948d8895-zflrs          1/1       Running     3          4d
default             pravega-pravega-segmentstore-0                       1/1       Running     0          4d
default             pravega-pravega-segmentstore-1                       1/1       Running     0          4d
default             pravega-pravega-segmentstore-2                       1/1       Running     0          4d
default             pravega-zk-0                                         1/1       Running     0          4d
default             pravega-zk-1                                         1/1       Running     0          4d
default             pravega-zk-2                                         1/1       Running     0          4d
default             pravega-zk-3                                         1/1       Running     0          4d
default             shaka-zulu-nfs-client-provisioner-59d7f8f84c-hfrsl   1/1       Running     1          4d
default             zookeeper-operator-685bfcbbc5-kzn46                  1/1       Running     0          4d
kube-system         heapster-85647cf566-zgj59                            1/1       Running     0          6d
kube-system         kube-dns-7559c96fc4-28r52                            3/3       Running     0          6d
kube-system         kubernetes-dashboard-5f4b59b97f-76j6r                1/1       Running     0          6d
kube-system         metrics-server-555d98886f-kzfhg                      1/1       Running     0          6d
kube-system         monitoring-influxdb-cdcf4674-q52hf                   1/1       Running     0          6d
kube-system         tiller-deploy-95d654d46-wxwrl                        1/1       Running     0          4d
pks-system          cert-generator-v0.11-85v64                           0/1       Completed   0          6d
pks-system          event-controller-56f9597df8-7hv8p                    2/2       Running     1          6d
pks-system          fluent-bit-2nwxj                                     2/2       Running     0          6d
pks-system          fluent-bit-58vfc                                     2/2       Running     0          6d
pks-system          fluent-bit-p8jdl                                     2/2       Running     0          6d
pks-system          fluent-bit-stnq2                                     2/2       Running     0          6d
pks-system          fluent-bit-tcghh                                     2/2       Running     0          6d
pks-system          sink-controller-65595c498b-frb5x                     1/1       Running     3          6d
pks-system          telemetry-agent-559f9c8855-pptwx                     1/1       Running     0          6d
pravega-longevity   small-manager-9468fbdf5-skgff                        1/1       Running     0          4d
pravega-longevity   small-tx-manager-6fb8856497-b77pp                    1/1       Running     0          4d
pravega-longevity   small-tx-worker-58779db76f-t2hrq                     1/1       Running     0          4d
pravega-longevity   small-worker-84879df646-cv9lw                        1/1       Running     0          4d
```

5. also ran benchmark with a new stream and its throwing the following exception :
```
./bin/pravega-benchmark -controller tcp://10.100.200.30:9090 -stream bench_again_23apr -events 100 -consumers 10 -producers 10 -time 120 -randomkey true
[main] WARN io.pravega.client.ClientConfig - The credentials are not specified or could not be extracted.
[main] WARN io.pravega.client.ClientConfig - The credentials are not specified or could not be extracted.
[main] INFO io.pravega.client.stream.impl.ControllerImpl - Controller client connecting to server at 10.100.200.30:9090
[main] WARN io.pravega.client.ClientConfig - The credentials are not specified or could not be extracted.
[main] WARN io.pravega.client.ClientConfig - The credentials are not specified or could not be extracted.
[main] INFO io.pravega.client.stream.impl.ControllerImpl - Controller client connecting to server at 10.100.200.30:9090
[main] INFO io.pravega.client.admin.impl.StreamManagerImpl - Creating scope: Scope
[StreamManager-Controller-1] INFO io.pravega.client.stream.impl.ControllerImpl - [requestId=7783168631765589747] Tagging client request (createScope-Scope).
[StreamManager-Controller-1] INFO io.pravega.client.stream.impl.ControllerResolverFactory - Updating client with controllers: [[[/10.100.200.30:9090]/{}]]


[grpc-default-executor-1] WARN io.pravega.client.stream.impl.ControllerImpl - [requestId=7783168631765589747] Failed to create scope: Scope
[grpc-default-executor-1] WARN io.pravega.client.stream.impl.ControllerImpl - [requestId=7783168631765589747] createScope failed: 
java.util.concurrent.CompletionException: io.pravega.client.stream.impl.ControllerFailureException: Failed to create scope: Scope
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
        at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:604)
        at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
        at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
        at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:632)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
        at io.pravega.client.stream.impl.ControllerImpl$RPCAsyncCallback.onCompleted(ControllerImpl.java:1026)
        at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:432)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:694)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:397)
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:459)
        at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:546)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:467)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:584)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: io.pravega.client.stream.impl.ControllerFailureException: Failed to create scope: Scope
        at io.pravega.client.stream.impl.ControllerImpl.lambda$createScope$1(ControllerImpl.java:208)
        at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
        ... 30 more
io.pravega.client.stream.impl.ControllerFailureException: Failed to create scope: Scope
        at io.pravega.client.stream.impl.ControllerImpl.lambda$createScope$1(ControllerImpl.java:208)
        at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
        at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
        at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
        at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:632)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
        at io.pravega.client.stream.impl.ControllerImpl$RPCAsyncCallback.onCompleted(ControllerImpl.java:1026)
        at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:432)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:694)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:397)
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:459)
        at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:546)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:467)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:584)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)

```

[pravega-k8s-logging-2019-04-18-15-23-06.zip](https://github.com/pravega/pravega/files/3106382/pravega-k8s-logging-2019-04-18-15-23-06.zip)

--- a/controller/src/main/java/io/pravega/controller/store/host/ZKHostStore.java
+++ b/controller/src/main/java/io/pravega/controller/store/host/ZKHostStore.java
@@ -83,7 +83,7 @@ public class ZKHostStore implements HostControllerStore {
             ZKUtils.createPathIfNotExists(zkClient, zkPath, HostContainerMap.EMPTY.toBytes());
             hostContainerMapNode.getListenable().addListener(this::updateMap);
             hostContainerMapNode.start(true);
-
+            updateMap();
             zkInit = true;
         }
     }

--- a/controller/src/test/java/io/pravega/controller/fault/SegmentContainerMonitorTest.java
+++ b/controller/src/test/java/io/pravega/controller/fault/SegmentContainerMonitorTest.java
@@ -146,7 +146,7 @@ public class SegmentContainerMonitorTest {
         monitor.startAsync().awaitRunning();
 
         assertEquals(hostStore.getContainerCount(), Config.HOST_STORE_CONTAINER_COUNT);
-        
+
         //Rebalance should be triggered for the very first attempt. Verify that no hosts are added to the store.
         assertTrue(sync.tryAcquire(10, TimeUnit.SECONDS));
         if (latches != null) {

--- a/controller/src/test/java/io/pravega/controller/fault/SegmentContainerMonitorTest.java
+++ b/controller/src/test/java/io/pravega/controller/fault/SegmentContainerMonitorTest.java
@@ -146,18 +146,13 @@ public class SegmentContainerMonitorTest {
         monitor.startAsync().awaitRunning();
 
         assertEquals(hostStore.getContainerCount(), Config.HOST_STORE_CONTAINER_COUNT);
-
-        assertEquals(0, hostStore.getHostContainersMap().size());
-        if (latches != null) {
-            latches.get(0).join();
-        }
-
+        
         //Rebalance should be triggered for the very first attempt. Verify that no hosts are added to the store.
         assertTrue(sync.tryAcquire(10, TimeUnit.SECONDS));
-        assertEquals(0, hostStore.getHostContainersMap().size());
         if (latches != null) {
             latches.get(1).join();
         }
+        assertEquals(0, hostStore.getHostContainersMap().size());
 
         //New host added.
         cluster.registerHost(new Host("localhost1", 1, null));

--- a/controller/src/test/java/io/pravega/controller/fault/SegmentContainerMonitorTest.java
+++ b/controller/src/test/java/io/pravega/controller/fault/SegmentContainerMonitorTest.java
@@ -147,18 +147,23 @@ public class SegmentContainerMonitorTest {
 
         assertEquals(hostStore.getContainerCount(), Config.HOST_STORE_CONTAINER_COUNT);
 
+        assertEquals(0, hostStore.getHostContainersMap().size());
+        if (latches != null) {
+            latches.get(0).join();
+        }
+
         //Rebalance should be triggered for the very first attempt. Verify that no hosts are added to the store.
         assertTrue(sync.tryAcquire(10, TimeUnit.SECONDS));
         assertEquals(0, hostStore.getHostContainersMap().size());
         if (latches != null) {
-            latches.get(0).join();
+            latches.get(1).join();
         }
 
         //New host added.
         cluster.registerHost(new Host("localhost1", 1, null));
         assertTrue(sync.tryAcquire(10, TimeUnit.SECONDS));
         if (latches != null) {
-            latches.get(1).join();
+            latches.get(2).join();
         }
         assertEquals(1, hostStore.getHostContainersMap().size());
 
@@ -169,7 +174,7 @@ public class SegmentContainerMonitorTest {
         cluster.deregisterHost(new Host("localhost1", 1, null));
         assertTrue(sync.tryAcquire(10, TimeUnit.SECONDS));
         if (latches != null) {
-            latches.get(2).join();
+            latches.get(3).join();
         }
         assertEquals(3, hostStore.getHostContainersMap().size());
 
@@ -181,7 +186,7 @@ public class SegmentContainerMonitorTest {
         //Wait for rebalance and verify the host update.
         assertTrue(sync.tryAcquire(10, TimeUnit.SECONDS));
         if (latches != null) {
-            latches.get(3).join();
+            latches.get(4).join();
         }
         assertEquals(4, hostStore.getHostContainersMap().size());
 

--- a/controller/src/test/java/io/pravega/controller/store/stream/HostStoreTest.java
+++ b/controller/src/test/java/io/pravega/controller/store/stream/HostStoreTest.java
@@ -29,8 +29,12 @@ import org.apache.curator.test.TestingServer;
 import org.junit.Assert;
 import org.junit.Test;
 
+import java.util.Map;
+import java.util.Set;
 import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
+
+import static org.junit.Assert.assertEquals;
 
 /**
  * Host store tests.
@@ -93,9 +97,16 @@ public class HostStoreTest {
                 latch.complete(null);
             });
             // Update host store map.
-            hostStore.updateHostContainersMap(HostMonitorConfigImpl.getHostContainerMap(host, controllerPort, containerCount));
+            Map<Host, Set<Integer>> hostContainerMap = HostMonitorConfigImpl.getHostContainerMap(host, controllerPort, containerCount);
+            hostStore.updateHostContainersMap(hostContainerMap);
             latch.join();
             validateStore(hostStore);
+
+            // verify that a new hostStore is initialized with map set by previous host store. 
+            HostControllerStore hostStore2 = HostStoreFactory.createStore(hostMonitorConfig, storeClient);
+
+            Map<Host, Set<Integer>> map = hostStore2.getHostContainersMap();
+            assertEquals(hostContainerMap, map);
         } catch (Exception e) {
             log.error("Unexpected error", e);
             Assert.fail();

